sdk/core/Azure.Core.Experimental/src/Auth/OAuth2BearerTokenAuthenticationPolicy.cs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Auth;
+
+/// <summary>
+/// A <see cref="PipelinePolicy"/> that uses an <see cref="ITokenProvider"/>
+/// </summary>
+/// <remarks>
+/// Initializes a new instance of the <see cref="OAuth2BearerTokenAuthenticationPolicy"/> class.
+/// </remarks>
+public class OAuth2BearerTokenAuthenticationPolicy : PipelinePolicy
+{
+    private readonly ITokenProvider _tokenProvider;
+    private readonly IScopedFlowToken _flowContext;
+
+    /// <param name="tokenProvider"></param>
+    /// <param name="contexts"></param>
+    public OAuth2BearerTokenAuthenticationPolicy(ITokenProvider tokenProvider, IEnumerable<IReadOnlyDictionary<string, object>> contexts)
+    {
+        _tokenProvider = tokenProvider;
+        _flowContext = GetContext(contexts, tokenProvider);
+    }
+
+    /// <inheritdoc />
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        Token token;
+        if (message.TryGetProperty(typeof(IScopedFlowToken), out var rawContext) && rawContext is IScopedFlowToken scopesContext)
```

Comment:
Generated clients will optionally attach operation specific scopes if they were defined in the spec